File: src/main/java/com/talkdesk/todolist/controller/TodoListController.java
package com.talkdesk.todolist.controller;

import com.talkdesk.todolist.model.TodoListItem;
import com.talkdesk.todolist.service.TodoListService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;


@RestController
@RequestMapping("/todo/{userId}")
public class TodoListController {

    private final TodoListService todoListService;

    public TodoListController(TodoListService todoListService) {
        this.todoListService = todoListService;
    }

    @GetMapping
    public Flux<TodoListItem> listItems(@PathVariable String userId) {
        return todoListService.getItemsForUser(userId);
    }

    @PostMapping
    public Mono<TodoListItem> createItem(@PathVariable String userId, @RequestBody Mono<CreateItemRequest> request) {
        return request.flatMap((data) -> todoListService.createItem(userId, data.getTitle(), data.getDescription()));
    }

    @PostMapping("/{id}")
    public Mono<ResponseEntity<TodoListItem>> completeItem(@PathVariable String userId, @PathVariable String id) {
        return todoListService.completeItem(userId, id)
                .map(ResponseEntity::ok)
                .defaultIfEmpty(ResponseEntity.notFound().build())
                .onErrorReturn(
                        TodoListService.ItemAlreadyCompletedException.class,
                        ResponseEntity.status(HttpStatus.FORBIDDEN).build()
                );
    }

    @DeleteMapping("/{id}")
    public Mono<ResponseEntity<Object>> deleteItem(@PathVariable String userId, @PathVariable String id) {
        return todoListService.deleteItem(userId, id)
                .map((item) -> ResponseEntity.noContent().build())
                .defaultIfEmpty(ResponseEntity.notFound().build());
    }

    public static class CreateItemRequest {

        private String title;
        private String description;

        public CreateItemRequest() {
        }

        public CreateItemRequest(String title, String description) {
            this.title = title;
            this.description = description;
        }

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }
    }

}

File: src/main/kotlin/com/talkdesk/todolist/controller/TodoListController.kt
package com.talkdesk.todolist.controller

import com.talkdesk.todolist.service.TodoListService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/todo/{userId}")
class TodoListController(
    private val todoListService: TodoListService
) {

    @GetMapping
    fun listItems(@PathVariable userId: String) = todoListService.getItemsForUser(userId)

    @PostMapping
    fun createItem(@PathVariable userId: String, @RequestBody request: Mono<CreateItemRequest>) =
        request.flatMap { data -> todoListService.createItem(userId, data.title, data.description) }

    data class CreateItemRequest(val title: String, val description: String)

    @PostMapping("/{id}")
    fun completeItem(@PathVariable userId: String, @PathVariable id: String) = todoListService.completeItem(userId, id)
        .map { item -> ResponseEntity.ok(item) }
        .defaultIfEmpty(ResponseEntity.notFound().build())
        .onErrorReturn(
            TodoListService.ItemAlreadyCompletedException::class.java,
            ResponseEntity.status(HttpStatus.FORBIDDEN).build()
        )

    @DeleteMapping("/{id}")
    fun deleteItem(@PathVariable userId: String, @PathVariable id: String) = todoListService.deleteItem(userId, id)
        .map { ResponseEntity.noContent().build<Any>() }
        .defaultIfEmpty(ResponseEntity.notFound().build())
}
